Ricardo Chávez Torres - 100406702

Bachelor in Data Science and Engineering

# Distribution of the covid-19 resources in Madrid

(first lets state that this situation is hypothetical)

There is a new treatment for covid19.
The deployment of this treatment requires qualified personal physician in a hospital that satisfies certain special conditions (not every hospital is valid)

Our objective is to deploy this treatment such as the most number of people is covered given the budget.
For this we hav to take into account that different hospitals, due to its geografical position and its installations, can treat a different number of cases with the same ammount of money, and that each hospital treats a different number of cases in each different region.

In order to distribute the resources fairly, each region must have the capability to treat at least 1% of its total population.

<img src="Img/Madrid density.png" alt="Madrid poblation" width="500" height="500"/>[1]


## a) Formulate the problem as a LP
### Sets

$\textbf{I} \equiv \ zones \ in \ Madrid \ ; \ \{Sur \ Metropolitano, ... , \ Sierra \ Sur\} $ 

$\textbf{J} \equiv\ Hospitals\ in\ which\ the\ cure\ can\ be\ deployed\ ;\{La\ Paz, ...,\ Puerta\ de\ Hierro \}$

### Parameters
$for \ \textbf{i} \in \textbf{I}$

$for \ \textbf{j} \in \textbf{J}$

$A_{i,j} \equiv\ Number\ of\ cases\ that\ can\ be\ treated\ in\ zone\ \textbf{i}\ by\ the\ hospital\ \textbf{j}\ per\ 1€\ spend\ in\ that\ hospital $

This table is $A$ but instead of the number of people treated by 1€ spent is by a 1000€ spent to facilitate the view

|                     |   Hospital la Paz |   Hospital Puerta de Hierro |   Hospital de Monteprincipe |   Hospital Santa Cristina |   Hospital de La Princesa |   Hospital Carlos III |   Hospital de Getafe |   Hospital del Henares |   Hospital Universitario Severo Ochoa |   Hospital de Guadarrama |   Hospital El Escorial |   Hospital Clínico San Carlos |   Hospital Virgen de la Poveda |
|:--------------------|------------------:|----------------------------:|----------------------------:|--------------------------:|--------------------------:|----------------------:|---------------------:|-----------------------:|--------------------------------------:|-------------------------:|-----------------------:|------------------------------:|-------------------------------:|
| Sur Metropolitano   |           130     |                      30     |                     110     |                     0     |                     0     |                 0     |                0     |                  0     |                                 0     |                    0     |                  0     |                         0     |                          0     |
| Este Metropolitano  |             0     |                     100     |                       0     |                     0     |                    42     |                 0     |                0     |                  0     |                                 0     |                    0     |                  0     |                         0     |                          0     |
| Oeste Metropolitano |            20     |                       4     |                      80     |                    25     |                     0     |                35     |                0     |                  0     |                                 0     |                    0     |                  0     |                         0     |                          0     |
| Norte Metropolitano |             0     |                       0     |                      30     |                     0     |                    30     |                35     |               36     |                  0     |                                 0     |                    0     |                  0     |                         0     |                          0     |
| Sierra Central      |             0     |                       0     |                      30     |                     0     |                     0     |                 0     |               35     |                 50     |                                 0     |                    0     |                  0     |                         0     |                          0     |
| Sudoeste Comunidad  |             0     |                       0     |                       0     |                     0     |                     0     |                 0     |                0     |                  0     |                                40     |                    0     |                  0     |                         0     |                          0     |
| Sudeste Comunidad   |             0     |                       0     |                       0     |                     0     |                     0     |                 0     |                0     |                  0     |                                15     |                   40     |                  0     |                         0     |                          0     |
| Nordeste Comunidad  |             0     |                       0     |                       0     |                     0     |                     0     |                 0     |                0     |                  0     |                                 0     |                    0     |                 25     |                         0     |                          0     |
| Sierra Norte        |             0     |                       0     |                       0     |                     0     |                     0     |                 0     |                0     |                  8     |                                 0     |                    0     |                  0     |                        20     |                          0     |
| Sierra Sur          |             0     |                       0     |                       0     |                     0     |                     0     |                 0     |                0     |                  0     |                                 0     |                    0     |                  0     |                         0     |                         20     |



$Budget_{j} \equiv\ limit\ budget(€)\ that\ a\ hospital\ \textbf{j}\ can\ expend.$


|                     |   Hospital la Paz |   Hospital Puerta de Hierro |   Hospital de Monteprincipe |   Hospital Santa Cristina |   Hospital de La Princesa |   Hospital Carlos III |   Hospital de Getafe |   Hospital del Henares |   Hospital Universitario Severo Ochoa |   Hospital de Guadarrama |   Hospital El Escorial |   Hospital Clínico San Carlos |   Hospital Virgen de la Poveda |
|:--------------------|------------------:|----------------------------:|----------------------------:|--------------------------:|--------------------------:|----------------------:|---------------------:|-----------------------:|--------------------------------------:|-------------------------:|-----------------------:|------------------------------:|-------------------------------:|
| Budget              |             1e+07 |                       1e+07 |                       1e+07 |                     1e+07 |                     1e+07 |                 1e+07 |                1e+07 |                  1e+07 |                                 1e+07 |                    1e+07 |                  1e+07 |                         1e+07 |                          1e+07 |

$Resources_{i} \equiv\ 1\%\ of\ the\ population\ of\ the\ region\ \textbf{i}$

| Sur Metropolitano   | Sur Metropolitano   | Este Metropolitano  | Oeste Metropolitano | Norte Metropolitano | Sierra Central      | Sudoeste Comunidad  | Sudeste Comunidad   | Nordeste Comunidad  | Sierra Norte        | Sierra Sur          |
|:--------------------|--------------------:|--------------------:|--------------------:|--------------------:|--------------------:|--------------------:|--------------------:|--------------------:|--------------------:|--------------------:|
|            Resources|            12922,40 |             6524,37 |             4720,43 |             3203,07 |              1677,36 |             1232,32 |             1023,62 |              594,29 |              407,48 |              344,99 |


### Variables

$x_j \equiv\ €\ spent\ in\ hospital\ \textbf{j}\ deploying\ the\ cure$

### Objective function
minimize the money spent

$min\sum_{j\in J} x_j $

### Constraints
$\sum_{j\in J} x_j A_{i,j} >= Resources_i; \forall\ \textbf{i}\in \textbf{I}$

$x_j  <= Budget_j; \forall\ \textbf{j}\in \textbf{J}$

In [1]:
from pyomo.environ import *
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from IPython.core.display import Image, display, Math, Latex

%matplotlib inline

plt.style.use('seaborn-talk')
plt.style.use('seaborn-colorblind')

In [2]:
url = "Data/Hospitales Madrid.csv"
data = pd.read_csv(url,index_col = 0)
display(data)

Hospital la Paz  Hospital Puerta de Hierro  \
Sur Metropolitano               0.13                      0.030   
Este Metropolitano              0.00                      0.100   
Oeste Metropolitano             0.02                      0.004   
Norte Metropolitano             0.00                      0.000   
Sierra Central                  0.00                      0.000   
Sudoeste Comunidad              0.00                      0.000   
Sudeste Comunidad               0.00                      0.000   
Nordeste Comunidad              0.00                      0.000   
Sierra Norte                    0.00                      0.000   
Sierra Sur                      0.00                      0.000   
Budget                    3000000.00                3500000.000   

                     Hospital de Monteprincipe  Hospital Santa Cristina  \
Sur Metropolitano                         0.11                      0.0   
Este Metropolitano                        0.00                      0.0   
Oeste Metropolitano                       0.08                      3.0   
Norte Metropolitano                       0.03                      0.0   
Sierra Central                            0.03                      0.0   
Sudoeste Comunidad                        0.00                      0.0   
Sudeste Comunidad                         0.00                      0.0   
Nordeste Comunidad                        0.00                      0.0   
Sierra Norte                              0.00                      0.0   
Sierra Sur                                0.00                      0.0   
Budget                              7300000.00                 500000.0   

                     Hospital de La Princesa  Hospital Carlos III  \
Sur Metropolitano                       0.00                0.000   
Este Metropolitano                      0.20                0.000   
Oeste Metropolitano                     0.00                0.035   
Norte Metropolitano                     0.03                0.035   
Sierra Central                          0.00                0.000   
Sudoeste Comunidad                      0.00                0.000   
Sudeste Comunidad                       0.00                0.000   
Nordeste Comunidad                      0.00                0.000   
Sierra Norte                            0.00                0.000   
Sierra Sur                              0.00                0.000   
Budget                            1700000.00          1000000.000   

                     Hospital de Getafe  Hospital del Henares  \
Sur Metropolitano                 0.000                 0.000   
Este Metropolitano                0.000                 0.000   
Oeste Metropolitano               0.000                 0.000   
Norte Metropolitano               0.036                 0.000   
Sierra Central                    0.035                 0.050   
Sudoeste Comunidad                0.000                 0.100   
Sudeste Comunidad                 0.000                 0.000   
Nordeste Comunidad                0.000                 0.000   
Sierra Norte                      0.000                 0.008   
Sierra Sur                        0.000                 0.000   
Budget                      1400000.000           1000000.000   

                     Hospital Universitario Severo Ochoa  \
Sur Metropolitano                                  0.000   
Este Metropolitano                                 0.000   
Oeste Metropolitano                                0.000   
Norte Metropolitano                                0.000   
Sierra Central                                     0.000   
Sudoeste Comunidad                                 0.040   
Sudeste Comunidad                                  0.015   
Nordeste Comunidad                                 0.000   
Sierra Norte                                       0.000   
Sierra Sur                                         0.000   
Budget                                       

## b) Implement the model

In [3]:
M = ConcreteModel()

I = list(data.iloc[:-1,:-1].index.values) 
J = list(data.iloc[:-1,:-1].columns.values)
M.I = Set(initialize=I)
M.J = Set(initialize=J)

A_data = data.iloc[:-1,:-1].stack().to_dict()
M.A = Param(M.I,M.J,initialize=A_data)

M.x = Var(M.J, within=NonNegativeReals,doc='Money spent in hospital j')

Resources_data = data.iloc[:-1,-1].to_dict()
M.Resources = Param(M.I,initialize=Resources_data,within=NonNegativeReals)

Budget_data = data.iloc[-1,:-1].to_dict()
M.Budget = Param(M.J, initialize=Budget_data,within=NonNegativeReals)

In [4]:
def objective_rule(M):
    return sum(M.x[j] for j in M.J)

M.obj = Objective(rule=objective_rule, sense=minimize, doc='deployment cost')

M.constraints = ConstraintList()
  
def Min_resources(i):
    return sum(M.x[j]* M.A[i,j] for j in M.J) >= M.Resources[i]

def Budget_lim (j):
    return M.x[j] <= M.Budget[j]

for i in I:
    M.constraints.add(sum(M.x[j]* M.A[i,j] for j in M.J) >= M.Resources[i])
    
for j in J:
    M.constraints.add(M.x[j] <= M.Budget[j])

In [5]:
Solver = SolverFactory('glpk')
M.dual = Suffix(direction=Suffix.IMPORT) 

Results = Solver.solve(M)
# M.x.display()
print('Total money expended ->', np.round(M.obj(),2),'€')

Total money expended -> 26353786.43 €


In [6]:
results_dataframe = pd.DataFrame({"Hospital":list(data.iloc[:-1,:-1].columns.values), 
             "Money Expended (€)": [round(M.x[j].value,2) for j in J]},index = ['' for j in J])
display(results_dataframe)

Hospital  Money Expended (€)
                      Hospital la Paz          3000000.00
            Hospital Puerta de Hierro          3308000.00
            Hospital de Monteprincipe          7300000.00
              Hospital Santa Cristina                0.00
              Hospital de La Princesa          1608185.00
                  Hospital Carlos III            76041.43
                   Hospital de Getafe          1400000.00
                 Hospital del Henares          1000000.00
  Hospital Universitario Severo Ochoa           580800.00
               Hospital de Guadarrama          2341250.00
                 Hospital El Escorial          2377160.00
          Hospital Clínico San Carlos          1637400.00
         Hospital Virgen de la Poveda          1724950.00

This solution tells us that to expend the least amount of money, we need to spend that much money in each hospital.

We can see that we won't give any money to Hospital Santa Cristina. This is because the regions that this hospital covers are already covered by another hospital but for less money.

## c) Compute the sensitivities

In [7]:
print('Sensitivities for minimum Resources constraints')
sensitivities_resources_dataframe = pd.DataFrame({"Region":list(data.iloc[:-1,:-1].index.values),
                                       "Population sensitivities (€/100 people vaiation)":np.round(list(dict(M.dual).values())[:len(I)],2)},
                                      index = ['' for i in I])
display(sensitivities_resources_dataframe)

print('\nSensitivities for Budget constraints')
sensitivities_budget_dataframe = pd.DataFrame({"Hospital":list(data.iloc[:-1,:-1].columns.values),
                                       "Budget sensitivities (€/Budget variation)":np.round(list(dict(M.dual).values())[len(I):],2)},
                                      index = ['' for j in J])
display(sensitivities_budget_dataframe)

Sensitivities for minimum Resources constraints


Region  Population sensitivities (€/100 people vaiation)
    Sur Metropolitano                                             30.95
   Este Metropolitano                                              0.71
  Oeste Metropolitano                                              0.00
  Norte Metropolitano                                             28.57
       Sierra Central                                              0.00
   Sudoeste Comunidad                                             15.62
    Sudeste Comunidad                                             25.00
   Nordeste Comunidad                                             40.00
         Sierra Norte                                             50.00
           Sierra Sur                                             50.00


Sensitivities for Budget constraints


Hospital  \
                      Hospital la Paz   
            Hospital Puerta de Hierro   
            Hospital de Monteprincipe   
              Hospital Santa Cristina   
              Hospital de La Princesa   
                  Hospital Carlos III   
                   Hospital de Getafe   
                 Hospital del Henares   
  Hospital Universitario Severo Ochoa   
               Hospital de Guadarrama   
                 Hospital El Escorial   
          Hospital Clínico San Carlos   
         Hospital Virgen de la Poveda   

  Budget sensitivities (€/Budget variation)  
                                      -3.02  
                                       0.00  
                                      -3.26  
                                       0.00  
                                       0.00  
                                       0.00  
                                      -0.03  
                                      -0.96  
                                       0.00  
                                       0.00  
                                       0.00  
                                       0.00  
                                       0.00

Each sensibility is corresponded with one constraint. The ones in the upper dataframe correspond with the constraint of how much people must be able to be treated in each region.

$\sum_{j\in J} x_j A_{i,j} >= Resources_i; \forall\ \textbf{i}\in \textbf{I}$

As $Resources_i = 1\% Population_i ; \forall i \in I$, each sensibility says how much should we spend if for the region $\textbf{i}$ if there were one more person to treat (100 more people in the region).

$sensibility_{Resources_i} \equiv \ the\ difference \ in \ the \ ammount\ of\ money\ that\ we\ should\ expend\ if\ the\ region\ \textbf{i}\ had\ 100\ more\ population$



For the sensibilities in the lower dataframe we can have a similar reading.
The constraint in this case is 

$x_j  <= Budget_j; \forall\ \textbf{j}\in \textbf{J}$

$sensibility_{Budget_j} \equiv \ the\ difference \ in \ the \ ammount\ of\ money\ that\ we\ should\ expend\ if\ the\ Hospital\ \textbf{j}\ had\ 1\ more\ euro\ of\ budget$

We can see that most hospitals have 0 sensibility. This is because they are not expending their whole Budget, so an increment in this would not make any difference.
We observe that the ones that have a non zero sensibility have a negative one.
This is because they have already spend all of their budget and by giving them more budget, they could cover the same regions that other hospitals are covering but without that much expense of money as they are more efficient. So, by increasing their budget the total money spend would be lower.

## d) Binary model

One more option has appear that might be more efficient in order to treat the virus.

This option is the construction of a temporary Hospital exclusively for the disease in each region.
This is interesting as the construction is very expensive, but the efficien of these new temporary hospitals is much higher than with the other hospitals.

There is also one problem with these hopsitals, to reduce the spread of the disease these specialized hospitals will only nurse the region in which they are.

These campaign hospitals have fixed cost to construct and fixed capability to attend.

### New parameters
$Ay = 100000 \equiv \ Number\ of\ cases\ that\ can\ be\ treated\ by\ a \ campaign \ hospital\ (fixed \ \forall \textbf{i} \in \textbf{I}) $

$Cy = 1000000€ \equiv \ Cost\ of\ deploying\ a\ campaign \ hospital (fixed \ \forall \textbf{i} \in \textbf{I}) $

### New varible

$y_i \equiv\ \{0,1\} wheter\ if \ a\ campaign\ hospital\ will\ be\ constructed\ in\ region\ \textbf{i} ; 0 \equiv no\ ; 1 \equiv yes$

### New objective function

$min\sum_{j\in J} x_j  + \sum_{i \in I} Cy · y_i$

### New restrictions 

$(\sum_{j\in J} x_j A_{i,j}) + y_i · Ay >= Resources_i; \forall\ \textbf{i}\in \textbf{I}$

$x_j  <= Budget_j; \forall\ \textbf{j}\in \textbf{J}$

In [8]:
M = ConcreteModel()

I = list(data.iloc[:-1,:-1].index.values) 
J = list(data.iloc[:-1,:-1].columns.values)
M.I = Set(initialize=I)
M.J = Set(initialize=J)

A_data = data.iloc[:-1,:-1].stack().to_dict()
M.A = Param(M.I,M.J,initialize=A_data)

M.x = Var(M.J, within=NonNegativeReals,doc='Money spent in hospital j')

M.y = Var(M.I, within=Binary,doc='binary indicating the construction or not of a campaign hospital in region i')

M.Ay = Param(initialize=100000)
M.Cy = Param(initialize=1000000)

Resources_data = data.iloc[:-1,-1].to_dict()
M.Resources = Param(M.I,initialize=Resources_data,within=NonNegativeReals)

Budget_data = data.iloc[-1,:-1].to_dict()
M.Budget = Param(M.J, initialize=Budget_data,within=NonNegativeReals)

In [9]:
def objective_rule(M):
    return sum(M.x[j] for j in M.J) + sum([M.y[i] * M.Cy for i in M.I])

M.obj = Objective(rule=objective_rule, sense=minimize, doc='deployment cost')

M.constraints = ConstraintList()
  
def Min_resources(i):
    return (sum(M.x[j]* M.A[i,j] for j in M.J) + M.y[i] * M.Ay)>= M.Resources[i]

def Budget_lim (j):
    return M.x[j] <= M.Budget[j]

for i in I:
    M.constraints.add((sum(M.x[j]* M.A[i,j] for j in M.J) + M.y[i] * M.Ay)>= M.Resources[i])
    
for j in J:
    M.constraints.add(M.x[j] <= M.Budget[j])

In [10]:
Solver = SolverFactory('glpk')
M.dual = Suffix(direction=Suffix.IMPORT) 

Results = Solver.solve(M)
# M.x.display()
# M.y.display()
print('Total money expended ->', np.round(M.obj(),2),'€')

Total money expended -> 21507555.0 €


In [11]:
results_dataframex = pd.DataFrame({"Hospital":list(data.iloc[:-1,:-1].columns.values), 
             "Money Expended (€)": [round(M.x[j].value,2) for j in J]},index = ['' for j in J])
display(results_dataframex)

results_dataframey = pd.DataFrame({"Region":list(data.iloc[:-1,:-1].index.values), 
             "Build a Hospital?": ['Yes' if round(M.y[i].value,2)==1 else 'No' for i in I]},index = ['' for i in I])
display(results_dataframey)

Hospital  Money Expended (€)
                      Hospital la Paz           3000000.0
            Hospital Puerta de Hierro           3308000.0
            Hospital de Monteprincipe           7300000.0
              Hospital Santa Cristina                 0.0
              Hospital de La Princesa           1608185.0
                  Hospital Carlos III                 0.0
                   Hospital de Getafe                 0.0
                 Hospital del Henares            232320.0
  Hospital Universitario Severo Ochoa                 0.0
               Hospital de Guadarrama             59050.0
                 Hospital El Escorial                 0.0
          Hospital Clínico San Carlos                 0.0
         Hospital Virgen de la Poveda                 0.0

Region Build a Hospital?
    Sur Metropolitano                No
   Este Metropolitano                No
  Oeste Metropolitano                No
  Norte Metropolitano               Yes
       Sierra Central                No
   Sudoeste Comunidad               Yes
    Sudeste Comunidad               Yes
   Nordeste Comunidad               Yes
         Sierra Norte               Yes
           Sierra Sur               Yes

We can see that by adding the possibility of the construction of these new hospitals the total money spent has decreased from 26353786€ to 21507555€. But the new hospitals is not always the way to go, it seems that for some regions, in order to cover the total resources needed at the lowest price it is better to not build the new hospital, like, for example in Sur Metropolitano or Este Metropolitano.



## d.2) Sensitivities
The built in functionality does not return the sensitivities for this model 

In [12]:
display(M.dual.display())

dual : Direction=Suffix.IMPORT, Datatype=Suffix.FLOAT
    Key : Value


None

so we are going to calculate them iteratively (as the problem seems to be easy enough so that our computer solves it very fast).


In [13]:
original_obj = np.copy(M.obj())
constraints_list = list(data.iloc[:-1,:-1].index.values) + list(data.iloc[:-1,:-1].columns.values)
sensitivities_binary_model = {}

In [14]:
for ii in constraints_list:
    M = ConcreteModel()

    I = list(data.iloc[:-1,:-1].index.values) 
    J = list(data.iloc[:-1,:-1].columns.values)
    M.I = Set(initialize=I)
    M.J = Set(initialize=J)

    A_data = data.iloc[:-1,:-1].stack().to_dict()
    M.A = Param(M.I,M.J,initialize=A_data)

    M.x = Var(M.J, within=NonNegativeReals,doc='Money spent in hospital j')

    M.y = Var(M.I, within=Binary,doc='binary indicating the construction or not of a campaign hospital in region i')

    M.Ay = Param(initialize=100000)
    M.Cy = Param(initialize=1000000)

    Resources_data = data.iloc[:-1,-1].to_dict()
    M.Resources = Param(M.I,initialize=Resources_data,within=NonNegativeReals)

    Budget_data = data.iloc[-1,:-1].to_dict()
    M.Budget = Param(M.J, initialize=Budget_data,within=NonNegativeReals)

    def objective_rule(M):
        return sum(M.x[j] for j in M.J) + sum([M.y[i] * M.Cy for i in M.I])

    M.obj = Objective(rule=objective_rule, sense=minimize, doc='deployment cost')

    M.constraints = ConstraintList()

    def Min_resources(i):
        return (sum(M.x[j]* M.A[i,j] for j in M.J) + M.y[i] * M.Ay)>= M.Resources[i]

    def Budget_lim (j):
        return M.x[j] <= M.Budget[j]

    for i in I:
        if ii == i:
            M.constraints.add((sum(M.x[j]* M.A[i,j] for j in M.J) + M.y[i] * M.Ay)>= M.Resources[i]+1)
        else:
            M.constraints.add((sum(M.x[j]* M.A[i,j] for j in M.J) + M.y[i] * M.Ay)>= M.Resources[i])

    for j in J:
        if ii == j:
            M.constraints.add(M.x[j] <= M.Budget[j]+1)
        else:
            M.constraints.add(M.x[j] <= M.Budget[j])

    Solver = SolverFactory('glpk')
    M.dual = Suffix(direction=Suffix.IMPORT) 

    Results = Solver.solve(M)
    # M.x.display()
    # M.y.display()
    sensitivities_binary_model[ii] = M.obj() - original_obj


In [15]:
sensitivities_binary_model_region = pd.DataFrame({'Region':list(sensitivities_binary_model.keys())[:len(I)],
                                                 'Population sensitivities binary model (€/100 people vaiation)': np.round(list(sensitivities_binary_model.values())[:len(I)],2)},
                                                 index = ['' for i in range(len(I))])

sensitivities_binary_model_hospital = pd.DataFrame({'Hospital':list(sensitivities_binary_model.keys())[len(I):],
                                                 'Budget sensitivities binary model (€/Budget variation)': np.round(list(sensitivities_binary_model.values())[len(I):],2)},
                                                 index = ['' for i in range(len(J))])
list(sensitivities_binary_model.values())[:len(I)]
list(sensitivities_binary_model.keys())[:len(I)]
display(sensitivities_binary_model_region)
display(sensitivities_binary_model_hospital)


Region  \
    Sur Metropolitano   
   Este Metropolitano   
  Oeste Metropolitano   
  Norte Metropolitano   
       Sierra Central   
   Sudoeste Comunidad   
    Sudeste Comunidad   
   Nordeste Comunidad   
         Sierra Norte   
           Sierra Sur   

  Population sensitivities binary model (€/100 people vaiation)  
                                              16.67              
                                               5.00              
                                               0.00              
                                               0.00              
                                               0.00              
                                              10.00              
                                              25.00              
                                               0.00              
                                               0.00              
                                               0.00

Hospital  \
                      Hospital la Paz   
            Hospital Puerta de Hierro   
            Hospital de Monteprincipe   
              Hospital Santa Cristina   
              Hospital de La Princesa   
                  Hospital Carlos III   
                   Hospital de Getafe   
                 Hospital del Henares   
  Hospital Universitario Severo Ochoa   
               Hospital de Guadarrama   
                 Hospital El Escorial   
          Hospital Clínico San Carlos   
         Hospital Virgen de la Poveda   

  Budget sensitivities binary model (€/Budget variation)  
                                              -1.17       
                                               0.00       
                                              -0.83       
                                               0.00       
                                               0.00       
                                               0.00       
                                               0.00       
                                               0.00       
                                               0.00       
                                               0.00       
                                               0.00       
                                               0.00       
                                               0.00

The interpretation for the sensitivity is very similar to the one of the previous model.
We can see in the first dataframe that when we increase the population of some regions the cost increases. And for the regions in which the cost does not increase is because it is already covered more than the sufficient ammount of resources.

In the second dataframe the values for the sensitivity are 0 when the hospital has not spend its whole budget and the sensitivity is negative when, if by increasing its budget, the total cost would decrease.

$sensibility_{Resources_i} \equiv \ the\ difference \ in \ the \ ammount\ of\ money\ that\ we\ should\ expend\ if\ the\ region\ \textbf{i}\ had\ 100\ more\ population$


$sensibility_{Budget_j} \equiv \ the\ difference \ in \ the \ ammount\ of\ money\ that\ we\ should\ expend\ if\ the\ Hospital\ \textbf{j}\ had\ 1\ more\ euro\ of\ budget$

[1] Image of population of madrid by statistical regions from *Instituto  de  Estadística  de  la  Comunidad  de  Madrid*.
http://www.madrid.org/iestadis/fijas/estructu/general/anuario/descarga/anumapzonas01.pdf

[2] Population data of Madrid by statistical regions from *Instituto  de  Estadística  de  la  Comunidad  de  Madrid*.
www.madrid.org/iestadis/fijas/estructu/demograficas/padron/descarga/padcmuniud.xls